src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,837 @@
+                            knowledge.push(predicate.map_to_global(&input_mapper, input_idx));
+                        }
+                    }
+
+                    let structured = PredicateStructure::new(&knowledge);
+                    let mut new_predicates = Vec::new();
+
+                    for equivalence in equivalences.iter_mut() {
+                        for expr in equivalence.iter_mut() {
+                            optimize(expr, &structured);
+                        }
+
+                        new_predicates.push(Constraint::Equivalence(equivalence.clone()));
+                    }
+
+                    knowledge.extend(new_predicates);
+                    knowledge
+                }
+                MirRelationExpr::Reduce {
+                    input,
+                    group_key,
+                    aggregates,
+                    ..
+                } => {
+                    let input_type = input.typ();
+                    let input_knowledge = self.action(input, let_knowledge)?;
+                    let structured = PredicateStructure::new(&input_knowledge);
+                    for key in group_key.iter_mut() {
+                        optimize(key, &structured);
+                    }
+                    for aggregate in aggregates.iter_mut() {
+                        optimize(&mut aggregate.expr, &structured);
+                    }
+
+                    // List of predicates we will return.
+                    let mut predicates = Vec::new();
+
+                    // Predicates that depend only on group keys should remain true of those columns.
+                    // 0. Form a map from key expressions to the columns they will become.
+                    let mut key_exprs = HashMap::new();
+                    for (index, expr) in group_key.iter().enumerate() {
+                        let key = expr.clone();
+                        let new_column = MirScalarExpr::Column(index);
+                        let existing_column =
+                            key_exprs.entry(key).or_insert_with(|| new_column.clone());
+                        // Report duplicated keys as equivalences
+                        predicates.push(Constraint::Equivalence(vec![
+                            existing_column.clone(),
+                            new_column,
+                        ]));
+                    }
+                    // 1. Visit each predicate, and collect those that reach no `ScalarExpr::Column` when substituting per `key_exprs`.
+                    //    They will be preserved and presented as output.
+                    for mut predicate in denormalize_predicates(input_knowledge.clone()) {
+                        if predicate.substitute(&key_exprs) {
+                            predicates.push(predicate);
+                        }
+                    }
+
+                    // TODO: Predicates about columns that are certain aggregates can be preserved.
```

Comment:
nit: given the code below, it's not clear what of the following comment is still "TODO".